### ISSUE TITLE
wrong result: TopK seems to use u8 comparisons when sorting values > u8::MAX

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#6333]
### What version of Materialize are you using?

```
v0.7.2-dev (375d3a5a7)
```

### What was the issue?

If ORDER BY + LIMIT is used in a materialized view, the result is incorrect for certain values of LIMIT for data values > 255.

### Is the issue reproducible? If so, please provide reproduction instructions.

```
CREATE TABLE t1 (f1 INTEGER);
INSERT INTO t1 VALUES (256);
INSERT INTO t1 VALUES (255);

CREATE MATERIALIZED VIEW v3 AS SELECT * FROM t1 ORDER BY f1 LIMIT 2;
CREATE VIEW
materialize=> SELECT * FROM v3;
 f1  
-----
 256 <- wrong result, 256 is sorted first
 255

```

The dataflow plan for the view looks needlessly complicated:

[Memory _ Materialize 0.7.2-dev.zip](https://github.com/MaterializeInc/materialize/files/6256996/Memory._.Materialize.0.7.2-dev.zip)

**Comment by philip-stoev** [comment]
```NUMERIC``` is also affected.

**Comment by frankmcsherry** [comment]
Just a quick comment. This isn't out of spec, as best as I understand it. The statement 
```sql
SELECT * FROM v3;
```
does not have an ordering statement, and so (according to SQL) has no ordering requirement in its output. The view `v3` certainly did order things, but it is not (according to SQL) wrong to discard that ordering in a new select query that does not state the ordering requirement.

I believe https://github.com/MaterializeInc/materialize/issues/724 is the issue tracking our thoughts here (that we might want to go beyond the spec to avoid confusion in cases like above). Of course, I'm just repeating what I understand the spec to say, so I could be totally wrong!

**Comment by philip-stoev** [comment]
I agree the spec does not demand ordering of views, but postgresql does return ordered results out ot views. For some views we do not even generate a TopK operator, e.g. #6332 . But I think the particular case above is worth investigating, there is obviously something that is happening on single bytes. Furthermore, the dataflow is very complicated for such a simple query.

**Comment by frankmcsherry** [comment]
At a single byte the `Row` ordering changes, which is what we use to sort results when we don't have an explicit requirement. The dataflow is complicated, but not needlessly (or at least, it is on purpose).

**Comment by philip-stoev** [comment]
In the EXPLAIN we declare that we sort on a particular column:

```
materialize=> explain view v3;
                 Optimized Plan                  
-------------------------------------------------
 %0 =                                           +
 | Get materialize.public.t1 (u49)              +
 | TopK group=() order=(#0 asc) limit=2 offset=0+
```

Anyway, here is an example query that returns a wrong result and that skirts the entire issue as to whether views should be ordered or not.

```
DROP SCHEMA public CASCADE;
CREATE SCHEMA PUBLIC;

CREATE TABLE t1 (f1 INTEGER);
INSERT INTO t1 VALUES (111);
INSERT INTO t1 VALUES (222);
INSERT INTO t1 VALUES (333);

SELECT (SELECT * FROM (SELECT f1 FROM t1 ORDER BY f1 LIMIT 3) AS a1 LIMIT 1);
 f1  
-----
 333
```
The correct result is ```111``` as confirmed by Postgres.

**Comment by frankmcsherry** [comment]
Only the `SELECT` with the `ORDER BY` statement has an ordering requirement. The `LIMIT 1` statement does not have an `ORDER BY` clause, and so the `LIMIT 1` is not required to use the ordering constraint of the `SELECT` subclause it contains. 

I'm going to tag in @benesch here for back up, because my strong feelings about SQL are limited.

**Comment by philip-stoev** [comment]
If we are not going to do the sorting then we should remove the ```order=(#0 asc) ``` from the EXPLAIN.

Here is a summary of what the user will see:
- only ORDER BY - discarded, not observed, not present in EXPLAIN
- ORDER BY + LIMIT 1 - the naively expected result is returned
- ORDER BY + LIMIT 2 over small integers - naively expected result is returned 
- ORDER BY + LIMIT 2 over larger integers - 50/50
- ORDER BY + LIMIT N over larger integers - not a chance

I think this is not internally consistent even if standards-compliant. I think we will be better off if we can get to the PostgreSQL behavior. If not, we should remove all mentions of ORDER BY in EXPLAIN for all cases.

**Comment by frankmcsherry** [comment]
I nominate that this issue (like #6332) be kicked up to SQL. The Optimization, Dataflow aspects of things are working as advertised, and afaict are all within spec. If it turns out that we want different behavior when people type `ORDER BY`, the SQL layer should make this call. #724 is the canonical instance of this issue.

**Comment by uce** [comment]
I'm also inclined to handle this (if at all) as part of #724. Moving this ticket to SQL/coordinator for now.

**Comment by philip-stoev** [comment]
Here is a test without any subqueries and  `LIMIT`s , to avoid them confusing the picture:
```
materialize=> create materialized view v1 as SELECT f1 FROM t1 ORDER BY f1;
CREATE VIEW
materialize=> SELECT * FROM v1;
 f1  
-----
 333
 111
 222
(3 rows)
```

The result is incorrect -- compare to running the same query without a view:

```
materialize=> SELECT f1 FROM t1 ORDER BY f1;
 f1  
-----
 111
 222
 333
(3 rows)
```

**Comment by frankmcsherry** [comment]
I think we still disagree on the SQL spec, which afaiu says that `SELECT * FROM v1` does not have an ordering requirement.

**Comment by frankmcsherry** [comment]
I really think this is a feature request, not a recognized bug. #724 is where that feature request is currently written down.

**Comment by benesch** [comment]
> I think we still disagree on the SQL spec, which afaiu says that SELECT * FROM v1 does not have an ordering requirement.

That is my understanding as well!
